### DISCUSSION TITLE
Rust?

### DISCUSSION THREAD

**Discussion by walterbender** [sugarlabs/musicblocks-v4#16] [General]
hmm could you do python --> wasm?
I don't know anything about wasm
ok so it might be too early
but there is https://github.com/RustPython/RustPython
so because compiling to wasm is an official target, one day you could run your python code in RustPython and generate WASM
should be an exciting decade for WASM too
a much needed web renaissance 
[ at work we run code on every continent, to make serving your Turtle Art pins fast, so ideally at some point we want to generate WASM out of C++... and run that as close to users as possible ]
Python directly running your browser:

https://rustpython.github.io/demo/
[ early days, but promising ] 
i could imagine the core algorithms of MB rely purely on the most fundamental parts of python (lists, dicts and that's it)
so I can access JS libraries? 
so maybe that would be libmusicblocks-core which could be then imported and wrapped 
yeah, but would the other way around be nicer?
from JS import the MB "virtual machine"
actually which ones are the juicy bits
the MB virt machine
I'd like to do everything in Python except the view
or the drawing
the view is the drawing and the webaudio
if the MB virt machine was in python, that'd mean nice unit tests plus easy to test in the console plus just generally nicer
is your code something like 30% drawing, 20% controller logic (wiring up view to events) and 50% the virt machine?
maybe even more VM than that
interesting, than having a toolchain that outputs reusable WASM might pay off nicely [ specially because of developer happiness -- you can crank out python without fighting the browser devel env ] 
*then having
ok so not too crazy, this project is doing it:

https://github.com/iodide-project/pyodide
if they got these packages properly running as WASM: 

NumPy, Pandas, Matplotlib, parts of SciPy, and NetworkX

getting the MB virt machine shouldn't be that much work...

**Comment by walterbender** [comment]
@meganindya -- this is an excerpt from a friend who is well versed in building scalable systems (for FB, Twitter, and Pintrest). Maybe we can take this approach instead? I for one would love to use Python.

**Comment by meganindya** [comment]
During initial discussions about the architecture this was one of the reasons I made the argument to entirely separate out the core; I thought the frontend application client should communicate with the core using a REST-like API.

JavaScript was invented for a specific purpose and with time its use grew unprecedentedly and we're here now with uncountable _npm_ packages. This growth was mostly accelerated by _node_. Nevertheless, it is not a performance oriented language. That's why _Dart_ came to my mind as it compiles to (faster) JavaScript or can be natively executed. But, _WASM_ with _C++_ or _Rust_ was the first thing I thought of. I did a bit of research but came to the conclusion that it is not there yet; probably, as of now, it'll be too much effort on our part as well.

But, I'd love to explore the possibility. I think it is imperative for us to extract as much juice as possible. Sure, we can optimize our algorithms multi-fold but there's no escaping from the fact that JavaScript is slow. And it is single threaded.

WASM does add the benefit of multiple language support on the browser natively, which is a a big plus. But, I guess the question still remains: are there enough resources currently for us to land this project on WASM? There's browser support issue as well.

**Comment by meganindya** [comment]
I found these:
https://madewithwebassembly.com/showcase/figma/
https://www.figma.com/blog/webassembly-cut-figmas-load-time-by-3x/

Looks promising.

**Comment by meganindya** [comment]
I did some reading and here are my observations:

1. _WASM_ will be indispensable in no time.
2. Code written in only a limited set of languages can be 'compiled' to _WASM_.
3. _WASM_ is based on _LLVM_, and therefore, code written in statically typed languages can be compiled to _WASM_.
4. _Python_ is dynamically typed, and has a _Garbage Collector_ which isn't currently supported by _WASM_.
5. _Python_ code can't be compiled to _WASM_; implementations like _RustPython_ actually compiled a _Python_ interpreter to _WASM_, so basically _Python_ runs in an interpreter which runs in _WASM_.
6. _C_/_C++_ is the best friend of _WASM_ currently, with close ones being _Rust_, _AssemblyScript_, _C#_, and _Go_.

I don't have enough information pertaining to the performance gain with using _Python_ in a _WASM_ compiled interpreter, given that it adds the bulk of carrying the interpreter with the application. With the languages mentioned in point 6, we'd actually get the benefit of _WASM_ without having to carry any additional weight.

**Reply by walterbender** [reply]
Let's discuss this with Raul on Wednesday. I'm not opposed to C++ but I would much prefer Python. But maybe it is a non-starter. In any case, I think you'll very much enjoy meeting him.

**Comment by walterbender** [comment]
And this: https://web.mit.edu/music21/
